### PR TITLE
Add TypeDocs for the NFT module

### LINKED ISSUE #9154: Add `NFT module`
### Description

Add NFT module reference for
- Summary of feature of the module
- module configs
- module constants
- module methods
- module endpoints

### Page Location
Typedoc

### PR DESCRIPTION
**Pull request by Tschakki**
### What was the problem?

This PR ...

- resolves #9154 
- restructures the export structure of the framework

### How was it tested?

`npx typedoc`
`open docs/typedoc/index.html`

https://liskhq.github.io/lisk-docs/lisk-sdk/references/typedoc/modules/lisk_framework.html

### CODE REVIEW

File: framework/src/modules/nft/endpoint.ts [hunk 1]
```diff
@@ -230,6 +302,18 @@ export class NFTEndpoint extends BaseEndpoint {
 		};
 	}
 
+	/**
+	 * Checks wheter a specific NFT is supported by the network.
```
Comment (TalhaMaliktz) [inline framework/src/modules/nft/endpoint.ts]:
```suggestion
	 * Checks whether a specific NFT is supported by the network.
```

File: framework/src/modules/base_module.ts [hunk 2]
```diff
@@ -15,46 +15,67 @@
...
 }
 
 export interface ModuleMetadata {
+	/** A list of Endpoints of the respective module. */
 	endpoints: {
+		// The name of the endpoint.
 		name: string;
+		// Required parameters for the endpoint.
 		request?: Schema;
+		// A schema of the expected response to a request to the endpoint.
```
Comment (TalhaMaliktz) [inline framework/src/modules/base_module.ts]:
```suggestion
		// A schema of the expected response to a request, sent to the endpoint.
```

File: framework/src/modules/nft/module.ts [hunk 3]
```diff
@@ -68,6 +68,41 @@ import {
 	MODULE_NAME_NFT,
 } from './constants';
 
+/**
+ * ## Description
+ * The `NFTModule` is used for creating, destroying NFTs (non-fungible tokens), and transferring them in the Lisk ecosystem.
+ *
+ * ## Not a stand-alone module
+ * The NFT module is not intended to be used as stand-alone module.
```
Comment (TalhaMaliktz) [inline framework/src/modules/nft/module.ts]:
```suggestion
 * The NFT module is not intended to be used as a stand-alone module.
```

File: framework/src/modules/nft/module.ts [hunk 4]
```diff
@@ -68,6 +68,41 @@ import {
...
+ * Instead, it should be used inside other modules, that intend to implement features related to NFTs.
+ * Other modules can use the provided {@link method | methods} of the NFT module, to implement custom commands for minting and destroying NFTs in the network.
+ * This allows to define the specific details about how NFT are created, and who is allowed to mint them.
```
Comment (TalhaMaliktz) [inline framework/src/modules/nft/module.ts]:
```suggestion
 * This allows us to define the specific details about how NFTs are created, and who is allowed to mint them.
```

File: framework/src/modules/nft/module.ts [hunk 5]
```diff
@@ -68,6 +68,41 @@ import {
...
+ *
+ * ## NFT Identifier
+ * To identify NFTs in the Lisk ecosystem, we introduce the `nftID`, a unique NFT identifier in the ecosystem.
+ * It is a 16 bytes long concatenation of the 4 bytes long `chainID`, the chain ID of the chain creating the NFT, the 4 bytes long `collectionID`, chosen when the NFT is created, and a 8 bytes long serialization of an index integer, automatically assigned at the NFT creation.
```
Comment (TalhaMaliktz) [inline framework/src/modules/nft/module.ts]:
```suggestion
 * It is a 16 bytes long concatenation of the 4 bytes long `chainID`, the chain ID of the chain creating the NFT, the 4 bytes long `collectionID`, chosen when the NFT is created, and an 8 bytes long serialization of an index integer, automatically assigned at the NFT creation.
```

File: framework/src/modules/nft/module.ts [hunk 6]
```diff
@@ -68,6 +68,41 @@ import {
...
+ *
+ * This allows chains to define multiple sets of NFTs, each identified by their respective collection. Each collection can then easily have its own attributes schema and custom logic.
+ * For example, an art NFT exchange could have a different collection per artist, index being then a unique integer associated with each art piece of this artist.
```
Comment (TalhaMaliktz) [inline framework/src/modules/nft/module.ts]:
```suggestion
 * For example, an art NFT exchange could have a different collection per artist, the index being a unique integer associated with each art piece of this artist.
```

File: framework/src/modules/nft/module.ts [hunk 7]
```diff
@@ -68,6 +68,41 @@ import {
...
+ *
+ * ## Attributes
+ * Each NFT is stored with an array of attributes specified by various modules, with each attribute property being a byte sequence that is not deserialized by the NFT module.
+ * Each custom module using NFTs should define schemas to serialize and deserialize their attributes property of NFTs.
+ *
+ * Note that the attributes properties are not limited in size by default, which can potentially cause the CCM {@link Modules.Interoperability.validateFormat} failure during the cross-chain NFT transfer.
+ *
+ * When an NFT is sent to another chain, the attributes properties of the NFT can be modified according to specifications set on the receiving chain.
```
Comment (TalhaMaliktz) [inline framework/src/modules/nft/module.ts]:
```suggestion
 * When an NFT is sent to another chain, the attributes and properties of the NFT can be modified according to specifications set on the receiving chain.
```

File: framework/src/modules/nft/module.ts [hunk 8]
```diff
@@ -68,6 +68,41 @@ import {
...
+ * When the NFT is received back on its native chain, the returned modified attributes are disregarded and the original attributes are restored, as currently defined by {@link https://github.com/LiskHQ/lips/blob/main/proposals/lip-0052.md#getnewattributes | getNewAttributes} function.
+ * If needed, custom modules can implement a more fine-grained approach towards the attributes that are modified cross-chain.
+ *
+ * ## NFT cross-chain transfers
+ * As is the case with fungible tokens, all NFTs are escrowed in the native chain.
+ * Technically, this means that when a token is sent from its native chain to another, the NFT is not erased, but the owner of the NFT now becomes the receiving chain.
+ * When an NFT is returned, the native chain can then check that the NFT is indeed coming from the chain it was sent to, and was not maliciously created and transferred from another chain.
```
Comment (TalhaMaliktz) [inline framework/src/modules/nft/module.ts]:
```suggestion
 * When an NFT is returned, the native chain can then check that the NFT is indeed coming from the chain it was sent to and was not maliciously created and transferred from another chain.
```

**Review by TalhaMaliktz — CHANGES_REQUESTED** [review]
I see a lot of changes in code files, what is the reason behind these changes? Are the changes a result of version update of the blockchain applications or is it something else?

File: framework/src/modules/nft/module.ts (hunk 4 above)
Comment (Tschakki) [reply framework/src/modules/nft/module.ts]:
```suggestion
 * This allows to define the specific details about how NFTs are created, and who is allowed to mint them.
```

File: framework/src/modules/nft/module.ts (hunk 7 above)
Comment (Tschakki) [reply framework/src/modules/nft/module.ts]:
```suggestion
 * When an NFT is sent to another chain, the `attributes` properties of the NFT can be modified according to specifications set on the receiving chain.
```

File: framework/src/controller/index.ts [hunk 9]
```diff
@@ -12,5 +12,6 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
-export { Controller } from './controller';
-export { InMemoryChannel, IPCChannel } from './channels';
+export * from './controller';
+export * from './channels';
+export * from './event';
```
Comment (ishantiw) [inline framework/src/controller/index.ts]: If we are not using other exports then we can avoid exposing them and especially the users will not use controller related stuff for their app as most of it is very internal to framework. We can revert this change

File: framework/src/engine/bft/index.ts [hunk 10]
```diff
@@ -12,7 +12,9 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
+export * from './constants';
 export { BFTModule } from './module';
 export type { BFTMethod } from './method';
+export * from './schemas';
```
Comment (ishantiw) [inline framework/src/engine/bft/index.ts]: schemas file has a lot of interfaces defined within it, it would be best to avoid exposing all of them but just expose `bftParametersSchema`, `BFTParameters` and `Validator`

File: framework/src/engine/consensus/index.ts [hunk 11]
```diff
@@ -13,13 +13,16 @@
...
-	CONSENSUS_EVENT_BLOCK_DELETE,
-	CONSENSUS_EVENT_BLOCK_NEW,
-} from './constants';
-export { BFTHeights } from '../bft/types';
+	BFTHeader,
+	ActiveValidator,
+	PkSigPair,
+	AggregateCommit,
+	CommitPool,
+	ValidatorUpdate,
```
Comment (ishantiw) [inline framework/src/engine/consensus/index.ts]: Probably some of them will never be used by user or not intended to be used by user
```suggestion
	BFTHeader,
	ActiveValidator,
	AggregateCommit,
```

File: framework/src/engine/endpoint/index.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+export * from './chain';
+export * from './consensus';
+export * from './state';
+export * from './system';
+export * from './txpool';
```
Comment (ishantiw) [inline framework/src/engine/endpoint/index.ts]: make sure we don't expose this to user as its internal and they only need to know about how to call these endpoints

File: framework/src/engine/index.ts [hunk 13]
```diff
@@ -12,5 +12,7 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
-export { Engine } from './engine';
-export { computeValidatorsHash, areDistinctHeadersContradicting, BFTParameters } from './bft';
+export * from './engine';
+export * from './bft';
+export * from './consensus';
+export * from './endpoint';
```
Comment (ishantiw) [inline framework/src/engine/index.ts]: We should only expose what is needed by a user or we handle it in src/index.ts file instead of exposing all of them

File: framework/src/modules/index.ts [hunk 14]
```diff
@@ -12,15 +12,23 @@
...
+export * as Token from './token';
+export * as NFT from './nft';
+export * as Interoperability from './interoperability';
+export * as PoS from './pos';
+export * as Reward from './reward';
+export * as DynamicReward from './dynamic_reward';
+export * as Random from './random';
+export * as PoA from './poa';
+export * as Validators from './validators';
+export * as Auth from './auth';
```
Comment (ishantiw) [inline framework/src/modules/index.ts:34]: i think we should only expose what is needed instead of exposing all from these files

**Review by ishantiw — CHANGES_REQUESTED** [review]
I think we need to restrict what we expose from framework. There are a lot of stuff which users will not use for their sidechain development as well as their are some exports that are internal to framework, meaning that it is only intended to be used within framework

File: framework/src/engine/bft/index.ts (hunk 10 above)
Comment (Tschakki) [reply framework/src/engine/bft/index.ts]: `Validator` is not exportable from here

File: framework/src/modules/index.ts (hunk 14 above)
Comment (Tschakki) [reply framework/src/modules/index.ts:34]: What is exposed for each module is defined in the `index.ts` of each module.

File: framework/src/engine/index.ts (hunk 13 above)
Comment (Tschakki) [reply framework/src/engine/index.ts]: We already reduce the exports in the `index.ts` files of engine, bft and consensus

File: framework/src/engine/consensus/index.ts [hunk 15]
```diff
@@ -13,13 +13,13 @@
  */
 
 export { Consensus } from './consensus';
+export * from './constants';
```
Comment (ishantiw) [inline framework/src/engine/consensus/index.ts]: we can revert back to the original exports from constants as we don't want to expose p2p rpc endpoints constants

File: framework/src/engine/consensus/index.ts [hunk 16]
```diff
@@ -13,13 +13,13 @@
...
-	CONSENSUS_EVENT_FORK_DETECTED,
-	CONSENSUS_EVENT_BLOCK_DELETE,
-	CONSENSUS_EVENT_BLOCK_NEW,
-} from './constants';
-export { BFTHeights } from '../bft/types';
+	BFTHeader,
+	ActiveValidator,
+	AggregateCommit,
+} from './types';
+export * from './schema';
```
Comment (ishantiw) [inline framework/src/engine/consensus/index.ts]: we don't expose any schema as its internal to consensus

File: framework/src/engine/bft/index.ts [hunk 17]
```diff
@@ -12,7 +12,9 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
+export * from './constants';
```
Comment (ishantiw) [inline framework/src/engine/bft/index.ts]: BFT constants are mostly used within BFT only and we can remove this

File: framework/src/engine/index.ts [hunk 18]
```diff
@@ -12,5 +12,6 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
-export { Engine } from './engine';
-export { computeValidatorsHash, areDistinctHeadersContradicting, BFTParameters } from './bft';
+export * from './engine';
```
Comment (ishantiw) [inline framework/src/engine/index.ts]: we only expose what we need from the file, so better to only expose `Engine` class only as previously done

**Review by ishantiw — CHANGES_REQUESTED** [review]
Few more comments around exports

**Review by TalhaMaliktz — APPROVED** [review]
LGTM 👍🏼

**Review by ishantiw — APPROVED** [review]
Great job @Tschakki 👏

**Review by has5aan — APPROVED** [review]

**Review by ishantiw — CHANGES_REQUESTED** [review]
We need to re-export the `types` specifically from `framework/src/index.ts` file instead of using `export *` otherwise its not exported from the package when its built.
Please revert all the changes related to re-exporting types and export them individually.

**Review by ishantiw — APPROVED** [review]
